### PR TITLE
Adding extra JsError safety checks

### PR DESCRIPTION
**Pull request by ssivov**

### CODE REVIEW

File: clarity.d.ts [hunk 1]
```diff
@@ -291,6 +291,7 @@ interface ITotalByteLimitExceededEventState extends IInstrumentationEventState {
 
 interface IClarityAssertFailedEventState extends IInstrumentationEventState {
   source: string;
+  comment: string;
```
Comment (toby-walker) [inline clarity.d.ts]: Bump version. :-)

File: src/core.ts [hunk 2]
```diff
@@ -116,8 +116,8 @@ export function instrument(eventState: IInstrumentationEventState) {
 }
 
 function getUnixTimestamp(): number {
-  return (window.performance && typeof performance.now === "function")
-    ? performance.now() + performance.timing.navigationStart
+  return (window.performance && window.performance.now)
```
Comment (toby-walker) [inline src/core.ts]: Why did you removed the typing check?

File: src/core.ts [hunk 3]
```diff
@@ -135,7 +135,7 @@ function envelope(): IEnvelope {
   return {
     clarityId: cid,
     impressionId,
-    url: top.location.href,
+    url: location.href,
```
Comment (toby-walker) [inline src/core.ts]: Would prefer window.location.href to be clearer.

Comment (toby-walker) [inline src/core.ts]: Why the change from window.top.location to window.location?

Comment (toby-walker) [inline src/core.ts]: Is window.location supported in all browsers? See:

image

From: https://developer.mozilla.org/en-US/docs/Web/API/Window/location

File: src/plugins/layout.ts [hunk 4]
```diff
@@ -68,6 +68,7 @@ export default class Layout implements IPlugin {
     let discoverTime = getTimestamp();
     traverseNodeTree(document, this.discoverNode.bind(this));
     this.shadowDomConsistent = this.shadowDom.mirrorsRealDom();
+    assert(this.shadowDomConsistent, "discoverDom", "shadowDom inconsistent after dom discovery");
```
Comment (toby-walker) [inline src/plugins/layout.ts]: CAn you return an indication of what inconsistency was? I would encourage you to think about a "diagnostic" approach rather than a simple "pass/fail" approach. You compute diagnostics that indicate issues and can then return what the issues found were and which diagnostics cause a failure. 

This is probably more than the scope of this PR but worth thinking about more.

File: src/plugins/layout/shadowdom.ts [hunk 5]
```diff
@@ -88,7 +92,11 @@ export class ShadowDom {
 
   public updateShadowNode(index: number, newLayout?: ILayoutState) {
     let shadowNode = this.getShadowNode(index);
-    assert(!!shadowNode, "updateShadowNode");
+
+    assert(!!shadowNode, "updateShadowNode", "shadowNode is missing");
```
Comment (toby-walker) [inline src/plugins/layout/shadowdom.ts]: Add more useful message to indicate what is missing. Add a way to show the element and its location in the tree and include that in the message.

File: src/plugins/viewport.ts [hunk 6]
```diff
@@ -1,8 +1,8 @@
 import { addEvent, bind } from "../core";
 
 export default class Viewport implements IPlugin {
-  private eventName : string = "Viewport";
-  private distanceThreshold : number = 20;
+  private eventName = "Viewport";
+  private distanceThreshold = 20;
```
Comment (toby-walker) [inline src/plugins/viewport.ts]: Low pri but consider moving this to config.

File: src/core.ts (hunk 2 above)
Comment (ssivov) [inline src/core.ts]: No single strong reason, but:
1. I feel like it's redundant. It can never be anything else but a function (unless manually overriden by someone)
2. I've looked at other projects and how they deal with performance safety checks - vast majority just checks that performance.now exists
3. We've seen errors such as 'object 'performance' has no method 'now'', however every single use of performance.now in our code base was wrapped in this condition. So while i'm not sure how this error is possible in the first place, I decided to tweak a condition and see if it makes a difference.

File: src/core.ts (hunk 3 above)
Comment (ssivov) [inline src/core.ts]: When the page is loaded in the iframe, 'top.location' will try to access parent page's url, which is forbidden by browser security policies (unless both pages are from the same origin), so it causes a javascript error.

File: src/plugins/viewport.ts (hunk 6 above)
Comment (ssivov) [inline src/plugins/viewport.ts]: I agree that this has to be in the config as well - we've discussed it on Sarvesh's PR too, but we'll make a separate PR for that.

---
In reply to: [119988354](https://github.com/Microsoft/clarity-js/pull/16#discussion_r119988354) [](ancestors = 119988354)

File: src/core.ts (hunk 3 above)
Comment (ssivov) [inline src/core.ts]: Here is the more detailed support list. I don't think we should worry about IE < 9 at this point:
image

File: src/plugins/layout/shadowdom.ts (hunk 5 above)
Comment (ssivov) [inline src/plugins/layout/shadowdom.ts]: Discussed this offline:
I totally agree that for proper debugging richer and more robust information is needed, so i'll keep working on that, but in a separate PR. For now, we'll go with a plain yes/no check to find the bottleneck of the observed problems.

---
In reply to: [119988311](https://github.com/Microsoft/clarity-js/pull/16#discussion_r119988311) [](ancestors = 119988311)

**Comment by toby-walker** [comment]
:shipit:
